core/src/unprocessed_packet_batches.rs

```diff
@@ -1,209 +1,353 @@
-        let (sig_len, sig_size) = decode_shortu16_len(&packet.data).ok()?;
-        let msg_start = sig_len
-            .checked_mul(size_of::<Signature>())
-            .and_then(|v| v.checked_add(sig_size))?;
-        let msg_end = packet.meta.size;
-        Some(&packet.data[msg_start..msg_end])
+    pub fn iter(&mut self) -> impl Iterator<Item = &DeserializedPacket> {
+        self.message_hash_to_transaction.values()
     }
 
-    /// Returns whether the given `PacketBatch` has any more remaining unprocessed
-    /// transactions
-    pub fn update_buffered_packets_with_new_unprocessed(
-        &mut self,
-        _original_unprocessed_indexes: &[usize],
-        new_unprocessed_indexes: &[usize],
-    ) -> bool {
-        let has_more_unprocessed_transactions = !new_unprocessed_indexes.is_empty();
-        if has_more_unprocessed_transactions {
-            self.unprocessed_packets
-                .retain(|index, _| new_unprocessed_indexes.contains(index));
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &mut DeserializedPacket> {
+        self.message_hash_to_transaction.iter_mut().map(|(_k, v)| v)
+    }
+
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut DeserializedPacket) -> bool,
+    {
+        // TODO: optimize this only when number of packets
+        // with oudated blockhash is high
+        self.packet_priority_queue.clear();
+        self.message_hash_to_transaction
+            .retain(|_k, deserialized_packet| {
+                let should_retain = f(deserialized_packet);
+                if should_retain {
+                    let priority_queue_entry =
+                        PacketPriorityQueueEntry::from_packet(deserialized_packet);
+                    self.packet_priority_queue.push(priority_queue_entry);
+                }
+                should_retain
+            })
+    }
+
+    pub fn len(&self) -> usize {
+        self.packet_priority_queue.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.packet_priority_queue.is_empty()
+    }
+
+    fn push_internal(&mut self, deserialized_packet: DeserializedPacket) {
+        let priority_queue_entry = PacketPriorityQueueEntry::from_packet(&deserialized_packet);
+
+        // Push into the priority queue
+        self.packet_priority_queue.push(priority_queue_entry);
+
+        // Keep track of the original packet in the tracking hashmap
+        self.message_hash_to_transaction
+            .insert(deserialized_packet.message_hash(), deserialized_packet);
+    }
+
+    /// Returns the popped minimum packet from the priority queue.
+    fn push_pop_min(&mut self, deserialized_packet: DeserializedPacket) -> DeserializedPacket {
+        let priority_queue_entry = PacketPriorityQueueEntry::from_packet(&deserialized_packet);
+
+        // Push into the priority queue
+        let popped_priority_queue_entry = self
+            .packet_priority_queue
+            .push_pop_min(priority_queue_entry);
```

Comment:
why not [`replace_min`](https://docs.rs/min-max-heap/1.3.0/min_max_heap/struct.MinMaxHeap.html#method.replace_min)?

Reply:
Subtle difference here based on the docs: https://docs.rs/min-max-heap/1.3.0/src/min_max_heap/lib.rs.html#229-232

`replace_min` will pop the min first then push the new element

`push_pop_min` will push the element, then pop the min

So if the new element you are pushing is the min, then `replace_min` will do the wrong thing here